src/wayland/wayland_base.cpp

```diff
@@ -151,3 +151,23 @@ void mw::internal_error_processing_request(wl_client* client, char const* method
         std::current_exception(),
         std::string() + "Exception processing " + method_name + " request");
 }
+
+void mw::tried_to_send_unsupported_event(wl_client* client, wl_resource* resource, char const* event, int required_version)
+{
+    wl_client_post_implementation_error(
+        client,
+        "Mir tried to send %s@%d.%s to object with version %d (requires version %d)",
+        wl_resource_get_class(resource),
+        wl_resource_get_id(resource),
+        event,
+        wl_resource_get_version(resource),
+        required_version);
+
+    log_error(
```

Comment:
Nit: Maybe `log_critical`? Getting here means that we've hit an application-breaking Mir logic bug.